sdks/java/core/src/main/java/org/apache/beam/sdk/util/SerializableUtils.java

```diff
@@ -79,16 +82,22 @@ public static Object deserializeFromByteArray(byte[] encodedValue,
   }
 
   public static <T extends Serializable> T ensureSerializable(T value) {
-    @SuppressWarnings("unchecked")
-    T copy = (T) deserializeFromByteArray(serializeToByteArray(value),
-        value.toString());
-    return copy;
+    return clone(value);
   }
 
   public static <T extends Serializable> T clone(T value) {
+    final Thread thread = Thread.currentThread();
+    final ClassLoader loader = value == null
+            ? thread.getContextClassLoader() : value.getClass().getClassLoader();
```

Comment:
For [io] I think it doesnt worth it, beam stack is already big enough IMHO, do you really want to add a dep for it?

About TCCL you dont have the choice since it is all about internals and respecting the caller context. How do you see it?

Reply:
Instead of adding a dependency, we can copy the source (add a comment attributing where it was copied from) since both projects are licensed Apache 2.0 instead of adding the dependency.

deserializeByteArray to me doesn't make sense to use the TCCL as there is no practical way to support it across a distributed execution environment. In reality, we are just trying to fix the clone method to use the classloader of the class that is being copied. Do you believe there is a lot of value in supporting the TCCL in deserializeByteArray?

Reply:
I copied code from apache tomee so we are good on that area.

I think we want to fix both, in particular to include the sanity check done at the beginning of beam.sdk.io.Read.Bounded#Bounded which would never work otherwise.

Reply:
`ensureSerializable` within `Read.Bounded#Bounded` is code during pipeline construction and not during pipeline execution. Users don't really have any support today for complex custom classloader configurations during execution across Spark/Flink/Dataflow/... so as long as we make sure that `clone` works then the user will be able to submit a pipeline to these different runners.

The user will still be responsible for correctly configuring the execution environment otherwise their classes won't be able to be deserialized and they will not have any control to set a TCCL based upon the current way that most data processing distributed execution environments work.

So lets use the copied ClassLoaderObjectInputStream and drop using the TCCL to pass information through and take my suggestion about having a private version of deserializeByteArray which takes a class loader.

Reply:
Not dure I get you, seems you justified my impl. You need to respect the TCCL while the pipeline is constructed (in a EE or OSGi server dor instance). Also doesnt cost much to beam so really think it is saner and avoid people to hack beam to be able to create pipelines.

Reply:
The clone code never honors the TCCL since we always use the classloader of `value` (TCCL for the `null` value doesn't matter).

Reply:
Guess it is fine for now since value loader should be good but not for ensureSerializable. Any issue here?

Reply:
Since your update modified `ensureSerializable` to use `clone`, my suggestion of dropping using the TCCL and using the ClassLoaderObjectInputStream with a private implementation of `deserializeByteArray` that takes a classloader then there isn't an issue. Otherwise I don't understand your last statement.

Reply:
Think i get it, can check it tomorrow and can work but only until the code triggers some classloading using the tccl so think it is still more secure this way, no?

Reply:
The code ignores the TCCL in `clone`/`ensureSerializable` and replaces it with your own choice of what the TCCL should be so your not honoring the users TCCL.

Reply:
Reworked a bit that logic, idea is to ensure the value class is available in the TCCL and if not switch to the value classloader. Seems it is way more robust this way.